IPP Feedback Banner - final PR

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
### IPP Feedback Banner - Project's final PR

This is a complete feature PR consisting of the following partial PRs, reviewed and merged to the this feature branch:
1. https://github.com/woocommerce/woocommerce-android/pull/8159
2. https://github.com/woocommerce/woocommerce-android/pull/8165
3. https://github.com/woocommerce/woocommerce-android/pull/8180
4. https://github.com/woocommerce/woocommerce-android/pull/8205
5. https://github.com/woocommerce/woocommerce-android/pull/8225
6. https://github.com/woocommerce/woocommerce-android/pull/8230

Closes: https://github.com/woocommerce/woocommerce-android/issues/8116
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
The goal of the project is to collect feedback from IPP users. We target three groups of merchants and show dedicated banners and feedback surveys.

- Specs for the banner title and message texts: p1674116096400209/1674112248.357019-slack-C0354HSNUJH

- More details in the project thread: pdfdoF-20T-p2

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
The banner should be visible on the orders page. The condition required to see the banner is to have COD enabled. 
If the user has ever completed IPP transactions banner type 2. or 3. (more than 10 transactions in the last 30 days) is shown. If they haven't made any IPP transactions banner 1. is shown.

### Images
The banner is placed on the Orders screen. Three banner types are shown in the table below.

| 1. IPP newbie users  | 2. IPP beginners | 3. IPP active users |
| ------------- | ------------- | ------------- | 
| ![Screenshot_20230123-113541](https://user-images.githubusercontent.com/4527432/214398300-5eba97a6-2d05-4383-8726-c17c592b59ff.png)  | ![Screenshot_20230123-113713](https://user-images.githubusercontent.com/4527432/214398338-c26c8eda-4ffb-4226-91bd-783a5a77d0da.png)  | ![Screenshot_20230123-113834](https://user-images.githubusercontent.com/4527432/214398361-83feda1b-cb76-4ebe-8477-188d8ca00dd0.png) |

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Implement storage for banner and survey events

samiuelson committed:
Extract logic of payments plugin detection

AnirudhBhat committed:
Merge pull request #8159 from woocommerce/8116/storing-survey-dismiss-completion-events

Storage for IPP feedback banner and survey events

samiuelson committed:
Implement use case for deciding if banner should be displayed

samiuelson committed:
Format line length

samiuelson committed:
Implement logic returning survey url for user group

samiuelson committed:
Clean up code in `ShouldShowFeedbackBanner`

samiuelson committed:
Clean up code in `GetIPPFeedbackSurveyUrl`

samiuelson committed:
Bump FluxC version

samiuelson committed:
Simplify `GetActivePaymentsPlugin::invoke` signature

Declare function as operator function.

samiuelson committed:
Improve `GetIPPFeedbackSurveyUrl` use case

1. Use case returns not non-null result, throwing exceptions when API returns error or null data.
2. Added additional unit test to verify behavior in case of negative transactions count.
3. Cleaned up the code around unit tests.

samiuelson committed:
Simplify use case invocations

samiuelson committed:
Revert file added by mistake

samiuelson committed:
Merge pull request #8165 from woocommerce/8116/banner-behavior-business-logic

IPP feedback banner business logic

samiuelson committed:
Show feedback banner only to stores in CA&US

samiuelson committed:
Add use cases for storing banner actions in prefs

samiuelson committed:
Update `GetIPPFeedbackSurveyUrl` to return banner messages

samiuelson committed:
Improve `ShouldShowFeedbackBanner` code

samiuelson committed:
Improve `GetIPPFeedbackBannerDataTest` and `ShouldShowFeedbackBannerTest` code style

samiuelson committed:
Clean up code

samiuelson committed:
Remove comments

samiuelson committed:
Update unit `GetIPPFeedbackBannerDataTest`

Check against android.R.string resource ids instead of their handles form `GetIPPFeedbackBannerData` class.

samiuelson committed:
Merge pull request #8180 from woocommerce/8116/banner-behavior-business-logic

Business logic updates for IPP feedback survey banner

samiuelson committed:
Update survey urls

samiuelson committed:
Show feedback banner in Orders screen

samiuelson committed:
Implement redirection to survey

* Opening survey in the existing component `FeedbackSurveyFragment`. Modified it to accept custom url argument.

* Prioritising IPP survey banner over Orders banner in Orders screen.

samiuelson committed:
Add feature flag

samiuelson committed:
Update texts on the banner

Added three custom titles for each user group

samiuelson committed:
Test if two banners are not shown together

samiuelson committed:
Enable `IPP_FEEDBACK_BANNER` feature flag

samiuelson committed:
Clean up banner layout

samiuelson committed:
Merge branch 'feature/8116-ipp-feedback-banner' into 8116/banner-ui

samiuelson committed:
Update user group detection rules

* Adjusted `ShouldShowFeedbackBanner` to return true for Stripe plugin users
* Refactored `GetIPPFeedbackBannerData` unit tests
* Adjusted user group detection rules in `GetIPPFeedbackBannerData` use case
* Updated survey urls

samiuelson committed:
Add campaign name property to `IPPFeedbackBanner`

campaignName is needed for event tracking

samiuelson committed:
Add campaign name property to `IPPFeedbackBanner`

campaignName is needed for event tracking

samiuelson committed:
Track banner events

samiuelson committed:
Rename `IPPBannerState` to `IPPSurveyFeedbackBannerState`

samiuelson committed:
Fix banner feature flag condition

samiuelson committed:
Rename `*IPPBanner*` to `*IPPFeedbackBanner*`

samiuelson committed:
Extract font dimen

samiuelson committed:
Fix typo

samiuelson committed:
Clean up the `refreshOrdersBannerVisibilityWithDelay` fun

samiuelson committed:
Rename values

samiuelson committed:
Refactor IPP banner test methods in `OrderListViewModelTest`

1. Fix names
2. Split methods testing multiple things into separate ones

samiuelson committed:
Add tests for view state and events in `OrderListViewModelTest`

samiuelson committed:
Hide IPP banner in landscape

samiuelson committed:
Fix formatting

samiuelson committed:
Hide IPP banner in landscape

samiuelson committed:
Clean up code

samiuelson committed:
Clean up code

samiuelson committed:
Merge branch `8116/banner-ui`

samiuelson committed:
Merge pull request #8205 from woocommerce/8116/banner-ui

UI of the IPP feedback survey banner

samiuelson committed:
Merge branch 'feature/8116-ipp-feedback-banner' into 8116/business-logic-adjustments

samiuelson committed:
Rename value

samiuelson committed:
Merge pull request #8225 from woocommerce/8116/business-logic-adjustments

IPP feedback banner: user group detection rules update

samiuelson committed:
Merge branch `feature/8116-ipp-feedback-banner`

rossanafmenezes committed:
Merge pull request #8230 from woocommerce/8116/events-tracking

IPP feedback banner: events tracking

samiuelson committed:
Merge branch 'trunk' into feature/8116-ipp-feedback-banner

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -161,6 +173,16 @@ class OrderListViewModel @Inject constructor(
+            if (shouldShowFeedbackBanner) {
+                val bannerData = getIPPFeedbackBannerData()
+                viewState = viewState.copy(ippFeedbackBannerState = IPPSurveyFeedbackBannerState.Visible(bannerData))
+                trackIPPBannerEvent(AnalyticsEvent.IPP_FEEDBACK_BANNER_SHOWN)
```
⚠️ I'm not seeing unit tests for any of the tracking related to the IPP feedback survey. We might want to add it. WDYT?

samiuelson replied:
I've added the unit tests for verifying correct analytics events and props 👍🏼

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt:
```diff
@@ -0,0 +1,152 @@
+
+    private fun requireSuccessfulTransactionsSummaryResponse(response: WooPayload<WCPaymentTransactionsSummaryResult>) {
+        if (response.isError || response.result == null) {
+            throw IllegalStateException("Failed to fetch transactions summary")
```
❓ I'm not sure if throwing an exception when the API call fails is the right way here. I just experienced this on my testing site. The transactions summary API was returning a 500 error and my app was crashing.

I'm thinking we should just log the error instead of throwing it in other places in this file as well. Can we log the error with the cause instead of crashing? I think won't letting the user use the app because of an API failure on the feedback banner is not a good UX since ipp feedback survey is not a must for users to use the app..

 WDYT?

samiuelson replied:
Good catch, I'll fix that.

samiuelson replied:
This is fixed now. In case of backend errors IPP feedback banner is not shown, and the "Create orders..." banner is shown as usual.

AnirudhBhat reviewed (commented):
Thanks for your work on this @samiuelson. Great work!

I've smoke tested the app and it seems to be working fine. I couldn't do thorough testing today because of a migraine. I'll be AFK tomorrow and will be back on Friday. But it will be late for your PR merge.

Since I've already reviewed major part of the PR, I've skimmed through the code and it looks good.

I'll leave it to @vanessamaartinstt for testing the app.

I've left a couple of questions to think about. Let me know what you feel about those.

samiuelson committed:
Merge branch 'trunk' into feature/8116-ipp-feedback-banner

samiuelson committed:
Fix scenario when backend returns error

In case backend returns error don't show the banner and log the API error instead of throwing an exception

samiuelson committed:
Add unit tests checking if correct events are tracked

Also extracts event keys and values to `AnalyticsTracker` class

samiuelson committed:
Fix scenario when no payment plugins are active

samiuelson committed:
Fix scenario when backend returns null response

samiuelson committed:
Update test for scenario when backend returns null response

samiuelson commented:
> Thanks for your work on this @samiuelson. Great work!
> 
> I've smoke tested the app and it seems to be working fine. I couldn't do thorough testing today because of a migraine. I'll be AFK tomorrow and will be back on Friday. But it will be late for your PR merge.
> 
> Since I've already reviewed major part of the PR, I've skimmed through the code and it looks good.
> 
> I'll leave it to @vanessamaartinstt for testing the app.
> 
> I've left a couple of questions to think about. Let me know what you feel about those.

Thank you @AnirudhBhat so much for testing the app and reviewing all the PRs!

I’ve fixed the issues you identified and added unit tests verifying if correct analytics events are tracked.

After manually testing different scenarios myself, I’ve also fixed the behavior in the case no active payments plugins are detected and when the backend (potentially) returns a successful but null response. 

Given this is the last day to merge the PR I’d suggest doing it today (26th of Jan). If there are any disqualifying issues found during the Mob testing we’ll still be able to fix them during the beta testing phase or disable the IPP banner using the feature flag before releasing it.

**List of scenarios fixed directly in this PR:**
1. Merchant has COD enabled but no payments plugin installed. 

In the above ☝🏼 case app was not showing the banner. Now: should show “ipp_not_user” banner.

2. Merchant has COD enabled but the backend returns an error if the payments plugin is not fully configured or (potentially) a null response from `transactions/summary` endpoint. 

In the above ☝🏼 case app was crashing. Now: should show “ipp_not_user” banner.
Basically, it was wrongly assumed that the backend response will always be successful. Now in case of an error response it is logged, the app is not crashing and the IPP banner is shown.

3. Added unit tests to `OrderListViewModelTest` verifying if correct analytics events are tracked when interacting with the banner.

samiuelson committed:
Clean up code

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerDataTest.kt:
```diff
@@ -257,20 +257,21 @@ class GetIPPFeedbackBannerDataTest : BaseUnitTest() {
 
     @Test
-    fun `given endpoint returns null response, then should throw exception`() = runBlocking {
+    fun `given endpoint returns null response, then should detect newbie`() = runBlocking {
```
🔍 Not related to this PR and not important, but just to spread knowledge about this. We have `testBlocking` method which lives in the `BaseUnitTest`. In this case it doesn't change anything, but in general is preferred to ensure consistency in the way we handle coroutines in tests.

samiuelson replied:
Switched to `testBlocking` in both `GetIPPFeedbackBannerDataTest` and `ShouldShowFeedbackBannerTest`.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt:
```diff
@@ -46,6 +46,9 @@ class MainActivityViewModel @Inject constructor(
         launch {
             featureAnnouncementRepository.getFeatureAnnouncements(fromCache = false)
         }
+        prefs.setIPPFeedbackSurveyCompleted(false)
```
⚠️ @samiuelson is this intentionally pushed into the repo? Seems like we'd be reseting the banner after each restart of the app, unless I'm missing something.

samiuelson replied:
My bad 😞 Removed!

malinajirka replied:
No worries, thanks!

samiuelson replied:
I've committed the wrong file with [ad72b7b](https://github.com/woocommerce/woocommerce-android/pull/8247/commits/ad72b7bcd79bb49f7f2dadb58d217051af93c7e0). It's reverted with [a534717](https://github.com/woocommerce/woocommerce-android/pull/8247/commits/a5347176603b10d03fe7a913817b26e93f06c952). 

The intended changes are committed with [0046ec7](https://github.com/woocommerce/woocommerce-android/pull/8247/commits/0046ec7c0bc9aee0b88ca395df906c57d8e3e816).

samiuelson committed:
Remove test code

samiuelson committed:
Switch from `runBlocking` to `testBlocking` function

samiuelson committed:
Clean up code

malinajirka reviewed (approved):
Thanks for working on this @samiuelson, great job! 👏 

I've tested several scenarios, including deactivating WCPay and toggling COD and all worked as expected 🏅.
